### ISSUE TITLE
Manual flash over-exposes the image

### ISSUE THREAD

**Issue by QuinnFreedman** [QuinnFreedman/Camera3#4]
As mentioned in #2 , `request.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_ON_ALWAYS_FLASH);` in a still capture request does not have the desired effect of taking an image with a flash. 

One solution is to use `FLASH_MODE_SINGLE` as in the example I gave here: https://github.com/QuinnFreedman/Camera3/issues/2#issuecomment-417065944 .

However, as mentioned here: https://github.com/QuinnFreedman/Camera3/issues/2#issuecomment-419039267 , that causes the camera to not take the flash into account when auto exposing the image, resulting in an over exposed image.

**Comment by QuinnFreedman** [comment]
One solution is to use manual exposure or ISO settings like this:

```java
public void takePicture() {
    CaptureRequestConfiguration config = new CaptureRequestConfiguration() {
        @Override
        public void configure(CaptureRequest.Builder request) {
            request.set(CaptureRequest.CONTROL_MODE, CaptureRequest.CONTROL_MODE_OFF)
            request.set(CaptureRequest.CONTROL_AE_MODE, CaptureRequest.CONTROL_AE_MODE_OFF)
            request.set(CaptureRequest.SENSOR_SENSITIVITY, prefs.iso)
            request.set(CaptureRequest.SENSOR_EXPOSURE_TIME, prefs.exposureTime)
        }
    };

    cameraManager.captureImage(captureSession, Camera3.PRECAPTURE_CONFIG_NONE, config);
}
```

However, I assume you might want auto-exposure. I have some ideas about how to address this but it might require a tweak to the library. I'll see what I can do about it

**Comment by mevinDhun** [comment]
Yes i am looking for auto-exposure flash. Waiting for the fix. Thanks

**Comment by QuinnFreedman** [comment]
So the issue here is that when you take a flash image, the camera has to pick at least an exposure and probably an iso as well before the picture is actually taken. So if you just flash when the image is captured, the camera will have no way to know what exposure it should use.

I had thought that if you set the `CONTROL_AE_MODE` in the preview before you take the picture to tell the camera that you were *going to* flash the image, then it would take that into account and make a guess to expose it darker. But, it seems like this is not the case.

What the system camera app seems to do is turn on the flash in torch mode first to light the scene, then trigger the auto-exposure so that everything adjusts to the new lighting. This also helps reduce red eye, but it does mean that there is a ~0.5 second delay before the picture is taken.

I haven't checked if I give you enough power to use this method in Camera3 (i.e. if setting a single pre-capture request to `FLASH_MODE_TORCH` will keep the flash on till you take the picture or if it will turn back off with the next frame) but I'll test that soon and if not I'll work on a workaround.

**Comment by QuinnFreedman** [comment]
If you wan't a really nice solution, you could monitor the light meter and/or do some image analysis on the preview, plus check the preview's auto-exposed settings each frame to make your own "auto-exposure", taking into account whether or not you were going to flash. But the fact that the system camera doesn't do this makes me think it would be hard to do well. Probably because it depends on how near your subject is and how reflective it is. You could try to use a depth cloud if your hardware supports it, but it would take some pretty serious CV to make a guess about the reflective properties of a surface. I don't know what the research is on that.

Some day down the road I might be interested in trying to build something like that in to this Camera3 but that's a lot more work than I want to do right now.

**Comment by mevinDhun** [comment]
>If you wan't a really nice solution, you could monitor the light meter and/or do some image analysis on the preview, plus check the preview's auto-exposed settings each frame to make your own "auto-exposure", taking into account whether or not you were going to flash. But the fact that the system camera doesn't do this makes me think it would be hard to do well. Probably because it depends on how near your subject is and how reflective it is. You could try to use a depth cloud if your hardware supports it, but it would take some pretty serious CV to make a guess about the reflective properties of a surface. I don't know what the research is on that.
>
This is way too complicated. I have already got everything working except the auto exposure flash. Are you trying to say that i currently do not have a way to do it with your library?

**Comment by QuinnFreedman** [comment]
So it turns out that  the solution to this is a lot simpler than I thought. All you have to do is set the preview to the same `CONTROL_AE_MODE` as the capture request (e.g. `CONTROL_AE_MODE_ON_ALWAYS_FLASH`).

As far as I can tell, this is essentially telling the camera that you plan to take a flash photo so it should pull down its exposure/iso for the actual capture request way more than it otherwise would. (Exactly what I was talking about doing manually because I didn't realize it had this built in)

This is a little awkward if you want to be able to switch the flash on/off because right now there is no way to update the preview so you have to restart the camera. The next feature update (which I've been working on for a long time) will give you the ability to live-update previews and video recordings, and may have a special feature just for this.

**Comment by QuinnFreedman** [comment]
I just now published a new version `0.4.3` (0a866b48e546cb3455502e2ef4bb578d2fc36675). It should be available whenever bintray finishes processing it.

I still have a bit of testing to do and right now it only works for preview, not for recording video, but that will probably be the next thing I do.

**Comment by QuinnFreedman** [comment]
So I guess the official answer for this is when flash is turned on (or set to auto) call 
```java
previewHandler.updateRequestConfig(new CaptureRequestConfiguration() {
    @Override
    public void configure(CaptureRequest.Builder request) {
        request.set(CaptureRequest.CONTROL_AE_MODE,
                    CaptureRequest.CONTROL_AE_MODE_ON_ALWAYS_FLASH);
    // or CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH
    }
});
```

And then use the same AE-flash mode in your capture request (don't use `FLASH_MODE`).

**Comment by l0stpenguin** [comment]
After updating to 0.4.3, i tried this and the flash is not triggered on google pixel:
```java
public void takePicture() {
        previewHandler.updateRequestConfig(new CaptureRequestConfiguration() {
            @Override
            public void configure(CaptureRequest.Builder request) {
                request.set(CaptureRequest.CONTROL_AE_MODE,
                        CaptureRequest.CONTROL_AE_MODE_ON_ALWAYS_FLASH);
                // or CaptureRequest.CONTROL_AE_MODE_ON_AUTO_FLASH
            }
        });

        CaptureRequestConfiguration config = new CaptureRequestConfiguration() {
            @Override
            public void configure(CaptureRequest.Builder request) {
                request.set(CaptureRequest.CONTROL_AE_MODE,
                        CaptureRequest.CONTROL_AE_MODE_ON_ALWAYS_FLASH);
            }
        };
        cameraManager.captureImage(captureSession, Camera3.PRECAPTURE_CONFIG_NONE, config);
    }
```
On my Xiaomi Redmi 3X, the flash is triggered but the image is dark even though i am in bright light.
![jpeg_20180927_170241_-1360723222](https://user-images.githubusercontent.com/17478561/46147930-e396ae80-c277-11e8-96c5-55462468b2fb.jpg)


Any idea of what could be wrong?

**Comment by QuinnFreedman** [comment]
Yeah I tested this on a Google Pixel right before I pushed and it worked perfectly but I just tried it again and  it randomly flashes like 1 in 10 times. My only thought is that it could be some kind of auto-flash even though it's set to always flash, but I couldn't get it to reliably flash even in the dark. Also, when it does go off, my preview goes black for a little while, which is another bug I should look at.

I don't thing this is a bug in anything I wrote because I'm getting the same result when I clone Google's camera2basic example and just change the `AUTO_FLASH` to `ALWAYS_FLASH` (which was working the other night). I've had this problem before, which is why I switched to using `FLASH_MODE` instead, which overexposes the image.

I'll look into this a little more and see if I can figure out what other apps do. I briefly went back to my original idea of just turning the torch on before starting auto-exposure so that it would be exposed correctly for a manual flash (which I'm guessing is what the builtin camera does) but it didn't quite work. I'll try to figure out why.

One really bizarre fix I found is if you set the preview to `FLASH_MODE_SINGLE` or `FLASH_MODE_TORCH`  but set the capture request to `CONTROL_AE_MODE_ON_ALWAYS_FLASH` it always flashes and exposes the image perfectly, but it does do a double-flash and over exposes the preview for a second. That seems like unintended behavior too so I'm not sure it will help your other phones.

**Comment by l0stpenguin** [comment]
I tried the _bizarre fix_ you mentioned and it triggers the flash. But the picture is being taken after the flash is turned off, so it has no effect on the resulting picture.

**Comment by l0stpenguin** [comment]
After playing with the settings, i discovered that iIf i clone the google sample camera2Basic and set the _setAutoFlash_ method to this:
```java
private void setAutoFlash(CaptureRequest.Builder requestBuilder) {
        requestBuilder.set(CaptureRequest.CONTROL_AE_MODE, CameraMetadata.CONTROL_AE_MODE_ON_ALWAYS_FLASH);
        requestBuilder.set(CaptureRequest.FLASH_MODE, CameraMetadata.FLASH_MODE_OFF);
 }
```
The flash is triggered properly with it being flashed while the picture is taken. If i use the same settings in your library, the flash is not triggered at all. Can you have a look?

**Comment by QuinnFreedman** [comment]
@mevinDhun That's very weird but maybe hopeful at least. I'll try to figure out why that's happening as soon as I get a chance.

**Comment by l0stpenguin** [comment]
@QuinnFreedman Did you have a chance to take a look at this issue?

**Comment by QuinnFreedman** [comment]
@mevinDhun I've been looking into this and getting some weird results.

On my Pixel 1 (Android 8.1.0), setting `CONTROL_AE_MODE_ON_ALWAYS_FLASH` and `FLASH_MODE_OFF` like you did in the camera2Basic demo *does* trigger the flash and the image does look normal but for me it's because the the flash doesn't seem to be timed right with the picture so it just doesn't do anything. If I go in a totally dark room and take a picture, the flash goes off but the image is still totally dark,

I'm guessing the first reason why that setting didn't work in camera3 was because in camera2Basic `setAutoFlash(..)` is called on *both* the capture request *and* the preview request. When I add those two lines to config for the capture request and the preview request, then I get the same behavior as camera2Basic for the first image (flash goes off and image looks fine, but it's not actually doing anything.)

But, doing this uncovered a very weird bug in my code I guess because after the first image is captured the preview goes black and then after taking another picture or doing anything else it comes back but it doesn't flash the images anymore. It seems that this is triggered by setting `CONTROL_AE_MODE_ON_ALWAYS_FLASH` in the **preview**, which also seems sufficient to get this whole incorrect flashing behavior. (Even if I never set `FLASH_MODE` and don't set anything for the actual capture, I get the same result). There are no errors or log messages at all so I'm not really sure why this is happening, but I'll look into it.

I am beginning to feel like this is some Android bug. I also might try updating Android to see if it goes away.
